vm-allocator/src/id.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2018 The Chromium OS Authors. All rights reserved.
+// Copyright © 2019 Intel Corporation
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+//
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+use std::fmt::{self, Display};
+use std::result;
+
+#[derive(Debug)]
+pub enum Error {
+    Overflow,
+    Duplicated,
+}
+
+impl Display for Error {
+    // This trait requires `fmt` with this exact signature.
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match self {
+            Overflow => write!(f, "Integer is overflow"),
+            Duplicated => write!(f, "Integer being allocated is duplicated"),
+        }
+    }
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+/// Manages allocating unsigned 32-bit number usage.
+/// Use `IdAllocator` whenever an unsigned 32-bit number needs to be allocated to different users.
+///
+/// # Examples
+///
+/// ```
+/// # use vm_allocator::IdAllocator;
+///   IdAllocator::new(1, std::u32::MAX).map(|mut p| {
+///       assert_eq!(p.allocate(Some(1)).unwrap(), 1);
+///       assert_eq!(p.allocate(Some(3)).unwrap(), 3);
+/// });
+/// ```
+#[derive(Debug)]
+pub struct IdAllocator {
+    start: u32,
+    end: u32,
+    used_map: Vec<u32>,
+}
+
+impl IdAllocator {
+    /// Creates a new `IdAllocator` for managing u32 usage.
+    /// * `start` - The starting number to manage.
+    /// * `end` - The ending number to manage.
+    /// * `used_map` - The used numbers ordered from lowest to highest.
+    pub fn new(start: u32, end: u32) -> Option<Self> {
+        Some(IdAllocator {
+            start,
+            end,
+            used_map: Vec::new(),
+        })
+    }
+
+    fn first_usable_number(&self) -> Option<u32> {
+        if self.used_map.is_empty() {
+            return Some(self.start);
+        }
+
+        let mut previous = self.start;
+
+        for iter in self.used_map.iter() {
+            // We know the subtraction could not be invalid.
+            if *iter > previous {
+                return Some(previous);
+            } else {
+                match iter.checked_add(1) {
+                    Some(p) => previous = p,
+                    None => return None,
+                }
+            }
+        }
+        if previous <= self.end {
+            Some(previous)
+        } else {
+            None
+        }
+    }
+
+    /// Allocates a number from the managed region. Returns `Ok(allocated_id)`
+    /// when successful, or Error indicates the failure reason.
+    pub fn allocate(&mut self, number: Option<u32>) -> Result<u32> {
```

Comment:
Please split the comments into summary and description. I would also skip the part where you say that it returns Ok on success. The interesting part is the Error path. Rust [documentation guidelines](https://doc.rust-lang.org/1.30.0/book/2018-edition/ch14-02-publishing-to-crates-io.html#commonly-used-sections) specify that when a function can return error, you can describe when does the error occur by having `#Errors` section.

You could write something like:
```rust
\\\ Allocates a unique number from the managed region.
\\\ 
\\\ # Arguments
\\\
\\\ *`number` - If the `Option` is `Some`, the function reserves the specified number.
\\\  
\\\ # Errors
\\\
\\\ When `number` has a value, it returns an error when the number is out of the defined interval or the
\\\ number was previously reserved in a previous call of the function. When the number is not specified,
\\\ it returns an error when there are no more numbers available in the interval.
```